server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4740,36 +4738,30 @@ public void syncSegmentsFromGivenRemoteSegmentStore(
             remoteDirectory.init();
             remoteStore.incRef();
         }
-        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments = sourceRemoteDirectory
-            .initializeToSpecificCommit(primaryTerm, commitGeneration)
-            .getMetadata();
+        RemoteSegmentMetadata remoteSegmentMetadata = sourceRemoteDirectory.initializeToSpecificCommit(primaryTerm, commitGeneration);
+        Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments = remoteSegmentMetadata.getMetadata();
         final Directory storeDirectory = store.directory();
         store.incRef();
 
         try {
-            String segmentsNFile = copySegmentFiles(
-                storeDirectory,
-                sourceRemoteDirectory,
-                remoteDirectory,
-                uploadedSegments,
-                overrideLocal
+            copySegmentFiles(storeDirectory, sourceRemoteDirectory, remoteDirectory, uploadedSegments, overrideLocal);
+
+            final SegmentInfos infosSnapshot = store.buildSegmentInfos(
+                remoteSegmentMetadata.getSegmentInfosBytes(),
+                remoteSegmentMetadata.getGeneration()
             );
-            if (segmentsNFile != null) {
-                try (
-                    ChecksumIndexInput indexInput = new BufferedChecksumIndexInput(
-                        storeDirectory.openInput(segmentsNFile, IOContext.DEFAULT)
-                    )
-                ) {
-                    SegmentInfos infosSnapshot = SegmentInfos.readCommit(store.directory(), indexInput, commitGeneration);
-                    long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
-                    if (remoteStore != null) {
-                        store.commitSegmentInfos(infosSnapshot, processedLocalCheckpoint, processedLocalCheckpoint);
-                    } else {
-                        store.directory().sync(infosSnapshot.files(true));
-                        store.directory().syncMetaData();
-                    }
+
+            long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
+            // delete any other commits, we want to start the engine only from a new commit made with the downloaded infos bytes.
+            // Extra segments will be wiped on engine open.
+            for (String file : List.of(store.directory().listAll())) {
+                if (file.startsWith(IndexFileNames.SEGMENTS)) {
+                    store.deleteQuiet(file);
                 }
             }
+            assert Arrays.stream(store.directory().listAll()).filter(f -> f.startsWith(IndexFileNames.SEGMENTS)).findAny().isEmpty()
+                : "There should not be any segments file in the dir";
+            store.commitSegmentInfos(infosSnapshot, processedLocalCheckpoint, processedLocalCheckpoint);
```

Comment:
This still needs some re-factoring as a lot of code is duplicated. Will get it fixed in next commit